### DISCUSSION TITLE
What does mean  / v, t -> self.button_pressed(v, t)

### DISCUSSION THREAD

**Discussion by dagi0627** [Beormund/Tasmota32-Multi-Zone-Heating-Controller#7] [General]
Hello i have question that what does mean ```v, t ->``` piece of code ```string.format("Button%d#Action", i), / v, t -> self.button_pressed(v, t)``` I am not quite understand it. Thank you.

**Comment by Beormund** [comment]
Which script and line number are you referring to?

**Reply by dagi0627** [reply]
Hi, in line 1358
```        api.add_rule(
            string.format(self.fmt, zone+1), 
            / v, t -> api.set_timer(0, / -> self.on_changed(v, t))
        )```

**Reply by Beormund** [reply]
https://github.com/Beormund/Tasmota32-Multi-Zone-Heating-Controller/blob/058a4a4e31b9d54358e03110ed7291b5d9a9788b/heating.be#L1356-L1362

So add_trigger is an instance method which creates a berry rule. The first argument to add_rule is the trigger that fires the rule - string.format creates the trigger 'Button1#Action' (if zone is zero for example). The second argument is a closure:  `/ v, t -> api.set_timer(0, / -> self.on_changed(v, t))`. When a berry rule fires, it calls the function specified. In this case, add_rule will call the api.set_timer function passing in two arguments (the value of the trigger and the trigger itself). The second argument to api.set_timer is also a closure which is the instance method on_changed. In berry, when you pass an instance method (rather than a function) as an argument, a closure is required.

The reason why we don't pass on_changed directly to add_rule is that we require a deferred asynchronous call to avoid a race condition (executing code from within a rule handler can cause race conditions e.g. rules firing rules). Basically, when the add_rule fires, self.on_changed is called immediately but asynchronously - hence the value 0 being passed to set_timer. In python the equivalent might be something like asyncio.create_task().

For more info on closures check out: https://github.com/berry-lang/berry/wiki/Chapter-5#closure. for more info on Tasmota berry rules check out https://tasmota.github.io/docs/Berry/#rules. And for more info on Tasmota berry timers check out: https://tasmota.github.io/docs/Berry/#timers.

**Reply by dagi0627** [reply]
Thank you very much. I have been studying berry. Now one question in line 1139 that 

https://github.com/Beormund/Tasmota32-Multi-Zone-Heating-Controller/blob/058a4a4e31b9d54358e03110ed7291b5d9a9788b/heating.be#L1141-L1152

Here are why 3 argument write in director[mode](self, zone, duration). Becuase all above functions (def boost(zone, secs), def advance(zone), def auto(zone) etc...,) have one or two arguments.

**Reply by Beormund** [reply]
Normally when you call an instance method, self is passed to the method implicitly (like Python). Here the method call is unusual in that it is invoked dynamically via indexing into a list of methods. The compiler does not pass self implicitly in this edge case so you have to provide it explicitly.
